File: xuri-rpc-java/src/main/java/com/xuri/rpc/RemoteProxyManager.java
package com.xuri.rpc;

import java.lang.ref.WeakReference;
import java.util.Map;
import java.util.Set;
import java.util.concurrent.ConcurrentHashMap;

/**
 * 远程代理管理器，管理从远程获取的代理对象
 * 对应TS中: export class RemoteProxyManager
 *
 * 使用WeakReference(对应TS WeakRef)，允许GC回收不再使用的远程代理
 */
public class RemoteProxyManager {

    // id → WeakReference<远程代理对象>
    private final Map<String, WeakReference<Object>> map = new ConcurrentHashMap<>();

    // Client → 该Client相关的代理ID集合
    private final Map<Client, Set<String>> clientMap = new ConcurrentHashMap<>();

    public void set(String id, Object proxy, Client client) {
        map.put(id, new WeakReference<>(proxy));
        clientMap.computeIfAbsent(client, k -> ConcurrentHashMap.newKeySet()).add(id);
    }

    /**
     * 不指定Client的set方法
     * 在TS中 transformArg 和 createRemoteProxy 中调用时未传client参数
     */
    public void set(String id, Object proxy) {
        map.put(id, new WeakReference<>(proxy));
    }

    public Object get(String id) {
        if (!map.containsKey(id)) {
            return null;
        }
        WeakReference<Object> ref = map.get(id);
        Object result = ref != null ? ref.get() : null;
        if (result == null) {
            map.remove(id);
            return null;
        }
        return result;
    }

    public Map<Client, Set<String>> getClientMap() {
        return clientMap;
    }

    public void removeFromClientMap(Client client) {
        clientMap.remove(client);
    }
}

File: xuri-rpc-java/src/main/kotlin/com/xuri/rpc/RemoteProxyManager.kt
package com.xuri.rpc

import java.lang.ref.WeakReference
import java.util.concurrent.ConcurrentHashMap

/**
 * 远程代理管理器，管理从远程获取的代理对象
 * 对应TS中: export class RemoteProxyManager
 *
 * 使用 WeakReference (对应TS WeakRef)，允许GC回收不再使用的远程代理
 */
class RemoteProxyManager {

    // id → WeakReference<远程代理对象>
    private val map = ConcurrentHashMap<String, WeakReference<Any>>()

    // Client → 该Client相关的代理ID集合
    val clientMap = ConcurrentHashMap<Client, MutableSet<String>>()

    fun set(id: String, proxy: Any, client: Client) {
        map[id] = WeakReference(proxy)
        clientMap.getOrPut(client) { ConcurrentHashMap.newKeySet() }.add(id)
    }

    /**
     * 不指定Client的set
     * TS中 transformArg/createRemoteProxy 调用时未传client参数
     */
    fun set(id: String, proxy: Any) {
        map[id] = WeakReference(proxy)
    }

    fun get(id: String): Any? {
        val ref = map[id] ?: return null
        val result = ref.get()
        if (result == null) {
            map.remove(id)
            return null
        }
        return result
    }

    fun removeFromClientMap(client: Client) {
        clientMap.remove(client)
    }
}
